new-log-viewer/src/services/decoders/JsonlDecoder/index.ts

```diff
@@ -0,0 +1,242 @@
+        useFilter: boolean,
+    ): Nullable<DecodeResultType[]> {
+        if (useFilter && null === this.#filteredLogEventMap) {
+            return null;
+        }
+
+        // Prevents typescript potential null warning.
+        const filteredLogEventIndices: number[] = this.#filteredLogEventMap as number[];
+
+        const length: number = useFilter ?
+            filteredLogEventIndices.length :
+            this.#logEvents.length;
+
+        if (0 > beginIdx || length < endIdx) {
+            return null;
+        }
+
+        const results: DecodeResultType[] = [];
+        for (let i = beginIdx; i < endIdx; i++) {
+            // Explicit cast since typescript thinks `#filteredLogEventIndices[filteredLogEventIdx]`
+            // can be undefined, but it shouldn't be since we performed a bounds check at the
+            // beginning of the method.
+            const logEventIdx: number = useFilter ?
+                (filteredLogEventIndices[i] as number) :
+                i;
+
+            results.push(this.#getDecodeResult(logEventIdx));
+        }
+
+        return results;
+    }
+
+    /**
+     * Retrieves log event using index then decodes into `DecodeResultType`.
+     *
+     * @param logEventIdx
+     * @return Decoded result.
+     */
+    #getDecodeResult = (logEventIdx: number): DecodeResultType => {
+        let timestamp: number;
+        let message: string;
+        let logLevel: LOG_LEVEL;
+
+        // eslint-disable-next-line no-warning-comments
+        // TODO We could probably optimize this to avoid checking `#invalidLogEventIdxToRawLine` on
+        // every iteration.
+        if (this.#invalidLogEventIdxToRawLine.has(logEventIdx)) {
+            timestamp = INVALID_TIMESTAMP_VALUE;
+            message = `${this.#invalidLogEventIdxToRawLine.get(logEventIdx)}\n`;
+            logLevel = LOG_LEVEL.NONE;
+        } else {
+            // Explicit cast since typescript thinks `#logEvents[filteredIdx]` can be undefined,
+            // but it shouldn't be since the index comes from a class-internal filter.
+            const logEvent = this.#logEvents[logEventIdx] as JsonLogEvent;
+            logLevel = logEvent.level;
+            message = this.#formatter.formatLogEvent(logEvent);
+            timestamp = logEvent.timestamp.valueOf();
+        }
+
+        return [
+            message,
+            timestamp,
+            logLevel,
+            logEventIdx + 1,
+        ];
+    };
+
+    /**
+     * Parses each line from the data array as a JSON object and buffers it internally. If a
+     * line cannot be parsed as a JSON object, an error is logged and the line added to an
+     * invalid map.
+     *
+     * NOTE: The data array(file) is freed after the very first run of this method.
+     */
+    #deserialize () {
+        if (null === this.#dataArray) {
+            return;
+        }
+
+        const text = JsonlDecoder.#textDecoder.decode(this.#dataArray);
+        let beginIdx: number = 0;
```

Comment:
```suggestion
        let beginIdx = 0;
```

Unless you think it's necessary for some reason?